### PR TITLE
Provide API to move allocation inside kind

### PR DESCRIPTION
**Pull request by michalbiesek**
### Description
PR introduce API related to reduced memory fragmentation

### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [x] Code compiles without errors
- [x] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [x] Created tests which will fail without the change (if possible)
- [x] Extended the README/documentation (if necessary)
- [x] All newly added files have proprietary license (if necessary)
- [x] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/311)

### CODE REVIEW

File: include/memkind.h [hunk 1]
```diff
@@ -419,6 +419,15 @@ void *memkind_realloc(memkind_t kind, void *ptr, size_t size);
...
 void memkind_free(memkind_t kind, void *ptr);
 
+///
+/// \brief Try to transfer allocation to reduce fragmentation
+/// \note STANDARD API
+/// \param kind specified memory kind
+/// \param ptr pointer to the allocated memory
+/// \return Pointer to newly trasnfered allocated memory
+///
+void *memkind_transfer_allocation(memkind_t kind, void *ptr);
```
Comment (pbalcer) [inline include/memkind.h]: I don't like the name, it implies the object is being moved to outside of the heap.

`memkind_defrag_relocate()` ?

File: man/memkind.3 [hunk 2]
```diff
@@ -313,14 +316,44 @@ is supported by TBB heap manager described in
 .B ENVIRONMENT
 section since Intel TBB 2019 Update 4.
 .PP
+.BR memkind_transfer_allocation ()
+implement allocation transfer inside specific
```
Comment (pbalcer) [inline man/memkind.3]: You must describe this better.
... relocates the object if doing so could improve overall external fragmentation of the heap ...

Something to that effect.

File: include/memkind.h (hunk 1 above)
Comment (michalbiesek) [reply include/memkind.h]: Done

File: man/memkind.3 (hunk 2 above)
Comment (michalbiesek) [reply man/memkind.3]: Done

File: man/memkind.3 [hunk 3]
```diff
@@ -313,14 +316,44 @@ is supported by TBB heap manager described in
 .B ENVIRONMENT
 section since Intel TBB 2019 Update 4.
 .PP
+.BR memkind_defrag_reallocate ()
+reallocates the object inside specific
+.IR "kind" ,
+doing so could improve overall external fragmentation of the heap. In case of failure function return
```
Comment (pbalcer) [inline man/memkind.3]: IF

This reallocates the object conditionally. The condition is that it's worth doing :P

Comment (michalbiesek) [reply man/memkind.3]: Done at least first part. Regarding second part I tried to rephrase it but as always I appreciate a feedback.

File: man/memkind.3 [hunk 4]
```diff
@@ -313,14 +316,45 @@ is supported by TBB heap manager described in
 .B ENVIRONMENT
 section since Intel TBB 2019 Update 4.
 .PP
+.BR memkind_defrag_reallocate ()
+reallocates the object conditonally inside specific
+.IR "kind" .
+Function determines itself if it's worthwhile to move allocation to reduce degree of external fragmentation of the heap.
```
Comment (pbalcer) [inline man/memkind.3]: remove itself, otherwise :lgtm:

File: include/memkind.h [hunk 5]
```diff
@@ -419,6 +419,15 @@ void *memkind_realloc(memkind_t kind, void *ptr, size_t size);
 ///
 void memkind_free(memkind_t kind, void *ptr);
 
+///
+/// \brief Try to reallocate allocation to reduce fragmentation
+/// \note STANDARD API
+/// \param kind specified memory kind
+/// \param ptr pointer to the allocated memory
+/// \return Pointer to newly trasnfered allocated memory
```
Comment (jschmieg) [inline include/memkind.h]: transferred

Comment (michalbiesek) [reply include/memkind.h]: Done

**Pull request merged**